fsharp-backend/src/LibBackend/StaticAssets.fs

```diff
@@ -107,8 +95,127 @@ let latestDeployHash (canvasID : CanvasID) : Task<Option<string>> =
+
+    let status, lastUpdate =
+      match read.instantOrNone "live_at" with
+      | Some datetime -> Deployed, datetime
+      | None -> Deploying, read.instant "created_at"
+
+    { deployHash = deployHash
+      url = url canvasName deployHash Short
+      status = status
+      lastUpdate = lastUpdate })
+
+let startStaticAssetDeploy
+  (user : Account.UserInfo)
+  (canvasID : CanvasID)
+  (canvasName : CanvasName.T)
+  : Task<StaticDeploy> =
+
+  let deployHash =
+    $"{canvasID}{System.DateTime.Now.ToString()}"
+    |> sha1digest
+    |> Base64.urlEncodeToString
+    |> String.removeSuffix "="
+    |> String.toLowercase
+    |> String.take 10
+
+  Sql.query
+    "INSERT INTO static_asset_deploys
+      (canvas_id, branch, deploy_hash, uploaded_by_account_id)
+    VALUES (@canvasID, @branch, @deployHash, @uploadedBy)
+    RETURNING created_at"
+  |> Sql.parameters [ "canvasID", Sql.uuid canvasID
+                      "branch", Sql.string branch
+                      "deployHash", Sql.string deployHash
+                      "uploadedBy", Sql.uuid user.id ]
+  |> Sql.executeRowAsync (fun reader -> reader.instant "created_at")
+  |> Task.map (fun lastUpdate ->
+    { deployHash = deployHash
+      url = url canvasName deployHash Short
+      lastUpdate = lastUpdate
+      status = Deploying })
+
+
+// TODO: what should happen if the deploy hash doesn't exist?
+// TODO: what if the deploy is already finished?
+let finishStaticAssetDeploy
+  (canvasID : CanvasID)
+  (canvasName : CanvasName.T)
+  (deployHash : string)
+  : Task<StaticDeploy> =
+  Sql.query
+    "UPDATE static_asset_deploys
+      SET live_at = NOW()
+      WHERE canvas_id = @canvasID AND deploy_hash = @deployHash
+      RETURNING live_at"
+  |> Sql.parameters [ "canvasID", Sql.uuid canvasID
+                      "deployHash", Sql.string deployHash ]
+  |> Sql.executeRowAsync (fun reader -> reader.instant "live_at")
+  |> Task.map (fun lastUpdate ->
+    { deployHash = deployHash
+      url = url canvasName deployHash Short
+      lastUpdate = lastUpdate
+      status = Deployed })
+
+
+/// <summary>Removes references to a canvas' static asset deploy</summary>
+/// <remarks>
+/// Since postgres doesn't have named transactions, we just delete the db
+/// record in question. For now, we're leaving files where they are; the right
+/// thing to do here would be to shell out to `gsutil -m rm -r` - leaving  this
+/// for a later round of 'garbage collection' work, in which we can query for
+/// files/dirs not known to the DB and delete them. TODO.
+/// </remarks>
+let deleteStaticAssetDeploy
+  (user : Account.UserInfo)
+  (canvasID : CanvasID)
+  (deployHash : string)
+  : Task<unit> =
+
+  // TODO the query here only allows someone to delete a deploy if they're the
+  // one who uploaded it (via `AND uploaded_by_account_id=@userId`). If more
+  // than one user are working within a canvas, shouldn't we allow anyone
+  // involved to delete the deploy?
```

Comment:
Outstanding question - what do you think @pbiggar? Personally I want to remove the `AND uploaded_by_account_id=@uploadedBy` filter, as anyone who has access to a canvas should be able to delete a deploy. (this filter was stolen by our OCaml code)